lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs

```diff
@@ -0,0 +1,280 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- |
+Copyright: © 2018-2022 IOHK
+License: Apache-2.0
+
+Data type 'TxHistory' for storing a set of transactions.
+Transactions are encoded "as" expressed in DB tables.
+
+-}
+module Cardano.Wallet.DB.Store.Transactions.Model
+    ( DeltaTxHistory (..)
+    , TxHistory
+    , TxHistoryF (TxHistoryF)
+    , TxRelationF (..)
+    , tokenCollateralOrd
+    , tokenOutOrd
+    , mkTxHistory
+    , TxRelationTxOutCtxK (..)
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.DB.Sqlite.Schema
+    ( TxCollateral (..)
+    , TxCollateralOut (..)
+    , TxCollateralOutToken (..)
+    , TxIn (..)
+    , TxOut (..)
+    , TxOutToken (..)
+    , TxWithdrawal (..)
+    )
+import Cardano.Wallet.DB.Sqlite.Types
+    ( TxId (TxId) )
+import Cardano.Wallet.Primitive.Types.RewardAccount
+    ( RewardAccount )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( AssetId (AssetId) )
+import Cardano.Wallet.Primitive.Types.TokenPolicy
+    ( TokenName, TokenPolicyId )
+import Cardano.Wallet.Primitive.Types.TokenQuantity
+    ( TokenQuantity )
+import Control.Arrow
+    ( (&&&) )
+import Data.Delta
+    ( Delta (..) )
+import Data.Foldable
+    ( fold )
+import Data.Generics.Internal.VL
+    ( view, (^.) )
+import Data.List
+    ( sortOn )
+import Data.Map.Strict
+    ( Map )
+import Data.Word
+    ( Word32 )
+import Fmt
+    ( Buildable (build) )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Types.Coin as W
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.Tx as W
+import qualified Data.Map.Strict as Map
+
+-- | A context that carries a TxOut together with its tokens
+-- (this will be needed in the future for the DB Layer).
+data WithTxOut a =
+    WithTxOut
+    { withTxOut_value :: a, withTxOut_context :: Maybe (TxOut, [TxOutToken]) }
```

Comment:
```suggestion
-- | Decorate a value, typically a transaction input or a collateral input,
-- with the corresponding transaction output.
data WithTxOut txin = WithTxOut
    { value :: txin
    , txOut :: W.TxOut
    }
```

The general concept which we discussed would involve `TxOut` from the wallet, rather than from the `DB.Schema`. 🤔

Reply:
The Maybe is necessary and here the encoding is at DB relations level. So we'll see later how to factor out things, when required.

Reply:
Ah. 🤔 I like the `WithTxOut` type when it's very general (as discussed), but I'm a bit "meh" if it's too specific to the DB layer. But yeah, let's merge it and revisit it later. 👍

Reply:
fixed